net context: write only the maximum payload net_pkt can hold

tbursztyka opened the pull request:
Attempt to fix #13211 

@aurel32 , @pfalcon I could not get internet access to qemu. I have an heavy setup on my work laptop (vpn, proxies etc...) and it's probably getting on the way. Could you check?

Commit:
net/pkt: Add a function to get the available buffer for payload

This will take into account the family and the protocol, as well as
existing buffer occupation, to return the available buffer space that
can be used for payload.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/context: Update length according to available buffer for payload

UDP was the only one doing the right thing. Let's generalize it
relevantly.

Fixes #13211

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pfalcon commented:
@tbursztyka: You don't need internet access to reproduce the issue reported by @aurel32 originally. The sample `dumb_http_server` against which the regression was reported, works locally between qemu and host. Given that the previous "default" test, echo_server is gone, I propose that we standardize on `dumb_http_server` being the default test. Using it is simple:

1. Build/run samples/net/sockets/dumb_http_server.
2. If so desired, use `telnet 192.0.2.1 8080` to manually make an HTTP request (`GET / HTTP/1.0\r\n\r\n`) to see all details of working.
3. Otherwise, use `curl http://192.0.2.1:8080/`
4. Finally, put it under some load with `ab -n1000 http://192.0.2.1:8080/` (watch for consistent response size and times in the report).

So, I tested it and indeed this patches fixes the issue. My output:
~~~
Server Software:        
Server Hostname:        192.0.2.1
Server Port:            8080

Document Path:          /
Document Length:        2122 bytes

Concurrency Level:      1
Time taken for tests:   211.819 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      21810000 bytes
HTML transferred:       21220000 bytes
Requests per second:    47.21 [#/sec] (mean)
Time per request:       21.182 [ms] (mean)
Time per request:       21.182 [ms] (mean, across all concurrent requests)
Transfer rate:          100.55 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        2    5   2.5      4      42
Processing:     9   16   3.9     16     120
Waiting:        6   11   3.1     10      81
Total:         12   21   5.3     20     147

Percentage of the requests served within a certain time (ms)
  50%     20
  66%     21
  75%     22
  80%     23
  90%     25
  95%     30
  98%     36
  99%     43
 100%    147 (longest request)
~~~

rveerama1 commented on include/net/net_pkt.h:
```diff
@@ -2241,6 +2241,21 @@ void net_pkt_append_buffer(struct net_pkt *pkt, struct net_buf *buffer);
+ * Note: Unlike net_pkt_available_buffer(), this will take into account the
+ *       headers space.
+ *
+ * @param pkt   The net_pkt which payload buffer availabality should
```
typo, "availability"

rveerama1 commented on include/net/net_pkt.h:
```diff
@@ -2241,6 +2241,21 @@ void net_pkt_append_buffer(struct net_pkt *pkt, struct net_buf *buffer);
+ *
+ * @return the amount of buffer available for payload
+ */
+size_t net_pkt_available_payload_buffer(struct net_pkt *pkt,
```
Better to use just "u16_t proto". As we started to have other protocol types.

tbursztyka replied:
Not in this case no, here it has to be a valid IPPROTO_* value (or 0).
Usage of u16_t in net_context is because of sock_raw, and net_pkt will never asses any header size on AF_PACKET (or AF_UNSPEC etc...).

rveerama1 replied:
ok

tbursztyka commented:
@pfalcon Actually socket based echo apps still exist (and are way much simpler than original echo app that were removed). I still use them for testing. But I agree, using dumb_http_server/big_http_download as an extra default for tcp would be nice. Thanks for trying the patches.

aurel32 commented:
@tbursztyka Thanks for the patch. It works fine using QEMU, however it fails to work with the SAM E70 Xplained board. There is an improvement though, it blocks after downloading 1393 bytes. I do not have my IMXRT1020 board at hand to give a try, I can do that tonight.

pfalcon commented:
I also only tested with qemu_x86. @aurel32: do you have any other networked hw besides sam_e70?

aurel32 commented:
> I also only tested with qemu_x86. @aurel32: do you have any other networked hw besides sam_e70?

Yes, I have a MIMXRT1020 board, but not at hand currently. I'll test tonight.

aurel32 commented:
I have just tried with USB networking (`CONFIG_USB_DEVICE_NETWORK_EEM=y`) and it also doesn't fix the issue. Small payload (using `#define USE_BIG_PAYLOAD 0`) still works fine.

tbursztyka commented:
Is this the same issue (-ENOBUFS context_sendto_new) or something else? 

Note there are a lot of issues around TCP. The new net_pkt API and packet flow has changed timings (it's much faster) and has revealed numerous concurrent access issues on net_context  and tcp objects. These last ones aren't protected actually (net_context got a mutex recently)

For instance, if you start echo server, and echo client and then stops echo client: when server finally closes the tcp connection it crashes.

aurel32 commented:
> Is this the same issue (-ENOBUFS context_sendto_new) or something else?

No from the application point of view, everything is fine, ie the 2181 bytes of payload passed to send() are send in two steps, first 1452 bytes on the first call to send() and 729 bytes on the second call.
Now looking at the traffic with tcpdump, only the first 1452 bytes packet actually arrives to the host.

pfalcon commented:
Tried to test on frdm_k64f, but it's broken: https://github.com/zephyrproject-rtos/zephyr/issues/13251 , https://github.com/zephyrproject-rtos/zephyr/pull/13250

aurel32 commented:
The second packet has the last frag with a zero len. It seems the GMAC driver isn't prepared for that. That said a quick and dirty fix to ignore that frag doesn't seems to fix the issue.

aurel32 commented:
Actually my fix was to quick and dirty. The problem is indeed that the GMAC driver doesn't handle the case of a frag with zero length. I'll try to work on a clean fix later today. Should I base it on PR #13167?

aurel32 reviewed (approved):
After fixing the GMAC driver, I confirm this patch works fine. I'll send a PR to fix the GMAC driver.

tbursztyka commented:
@aurel32 nice. About USB, I think the MTU there is just wrong: 1522? That would explain why nothing goes through.

aurel32 commented:
A quick look at the code shows that at least `drivers/eth_dw.c` also needs a fix to handle zero len frags. I am not sure about `drivers/wifi/winc1500/wifi_winc1500.c`, someone might need to have a look. Same for the `ieee802154.c` and `openthread.c` L2 code.

tbursztyka commented:
There are indeed places that still play directly with "frags", which is not nice. We might have to provide net_pkt.h helpers to do that. So the code going through the net_buf list would be at a unique place instead of scattered all around, reducing this kind of issues. That said, I wonder why there is a frag with len 0. Over allocation? Something worth to look at.

aurel32 commented:
> @aurel32 nice. About USB, I think the MTU there is just wrong: 1522? That would explain why nothing goes through.

I confirm that reducing `NETUSB_MTU` to 1500 fixes the issue. Now I wonder why it got increased to this value in 766101cf097f1e344f4516bc2a32937dba702746, ie a few days ago.

tbursztyka commented:
Indeed, that's a mix up of "physical MTU" and network MTU.

pfalcon commented:
@aurel32, @tbursztyka: +1 for the hunch that there should not be fragments of size 0 fed into drivers. I hope @tbursztyka could look into it, perhaps he might need more info on how to reproduce it?

aurel32 commented:
> I hope @tbursztyka could look into it, perhaps he might need more info on how to reproduce it?

For me it's reproducible with the default dump_http_server sample. For example I configure it with:

`cmake -GNinja -D BOARD=mps2_an385 -DCONF_FILE="prj.conf overlay-smsc911x.conf" ..`

Then add the following debug code:
```
diff --git a/subsys/net/l2/ethernet/ethernet.c b/subsys/net/l2/ethernet/ethernet.c
index 408ba50a83..f8aec043cd 100644
--- a/subsys/net/l2/ethernet/ethernet.c
+++ b/subsys/net/l2/ethernet/ethernet.c
@@ -511,6 +511,7 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
        struct ethernet_context *ctx = net_if_l2_data(iface);
        u16_t ptype;
        int ret;
+       struct net_buf *frag;
 
        if (IS_ENABLED(CONFIG_NET_IPV4) &&
            net_pkt_family(pkt) == AF_INET) {
@@ -579,6 +580,12 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
        net_pkt_cursor_init(pkt);
 
 send:
+       frag = pkt->frags;
+       while (frag) {
+               NET_DBG("frag->len = %i", frag->len);
+               frag = frag->frags;
+       }
+
        ret = api->send(net_if_get_device(iface), pkt);
        if (ret != 0) {
                eth_stats_update_errors_tx(iface);
```

That should give you a log like that for the second packet:

```
[00:00:13.262,562] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 14
[00:00:13.262,565] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 40
[00:00:13.262,572] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 128
[00:00:13.262,575] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 128
[00:00:13.262,585] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 128
[00:00:13.262,588] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 128
[00:00:13.262,593] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 128
[00:00:13.262,599] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 89
[00:00:13.262,606] <dbg> net_ethernet.ethernet_send: (0x20000b84): frag->len = 0
```

tbursztyka commented:
@aurel32 thanks, now I know why. It's actually "nominal" as the allocator computes headers size on the maximum size it can take (On TCP that is). And because of fixed data size of buffer, it might be even for 1 byte of extra space computed, a net_buf will be allocated but never used.

I'll come with a fix for it, in this PR. Imo, it should not be up to each and every driver to handle this. net_pkt buffer should be trimmed at relevant place.

Commit:
net/pkt: Add a function to trim the buffer

pkt->buffer is represented by 1+ net_buf. If some are unused, this will
deallocates them.

This situation can happen on TCP where net_pkt allocator evaluates the
header size to its maximum size. Which space might not be (fully) used
in the end. On fixed data size buffer, this might end up by having last
buffer(s) not bein used. So better removing those.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@aurel32 I believe the added commit should fix the !frag->len issue. I need to test it more thoroughly,  however

aurel32 commented:
> @aurel32 I believe the added commit should fix the !frag->len issue. I need to test it more thoroughly, however

Thanks for the additional patch, I confirm that it's possible to send big TCP payload with the GMAC driver again.

mike-scott commented:
FYI, this patch also fixes DTLS communication over sockets.  I hadn't noticed it was broken till after the socket API changes were merged and have been debugging it for a few days.

jukkar reviewed (approved):
Nice work!
